Change handling of links to preserve ordering

rubenarakelyan opened the pull request:
**NOTE** To be reviewed by the Publishing API team.

Currently, the `PATCH link_sets` command works out the difference between the provided links payload and the links currently in the database for the link set concerned. Links are then created and/or deleted in the database to match the payload. This process does not preserve ordering, so new links are appended at the end of the link set.

This commit changes the command to delete all existing links in the database for the link set, and re-add all the links from the payload, thereby preserving the natural ordering in the payload. In addition, the ordering is also recorded in a new `position` field in the `links` table, to allow the ordering algorithm to be changed in future without affecting natural ordering.

All existing links in the database are given a `position` of zero to reflect the fact that they have no inherent order.

Trello: https://trello.com/c/LlO4fTIS/153-make-publishing-api-order-the-links-size-3-mazz

cc @kevindew

kevindew commented on db/migrate/20161003145646_add_position_to_links.rb:
```diff
@@ -0,0 +1,5 @@
+class AddPositionToLinks < ActiveRecord::Migration
+  def change
+    add_column :links, :position, :integer
```
If we were to set this as non-null with a default of 0 could we drop the other migration?

rubenarakelyan replied:
Too right, done

kevindew commented:
Looks good, nicely done 👍 

I'll have a chat with our team about the idea of using activerecord-import and where we collectively stand on less queries vs validations/callbacks running.

rubenarakelyan commented:
@kevindew I think both of those can be done to some extent:

* Validations can be run by amending the import code (https://github.com/zdennis/activerecord-import/wiki/Examples)

* Callbacks can be run *if* we get an in-memory collection of ActiveRecord objects (https://github.com/zdennis/activerecord-import/wiki/Callbacks)

rubenarakelyan commented on app/commands/v2/patch_link_set.rb:
```diff
@@ -18,17 +18,25 @@ def call
 
-          content_ids_to_create = payload_content_ids - existing_content_ids
-          content_ids_to_delete = existing_content_ids - payload_content_ids
+          links.where(target_content_id: existing_content_ids).delete_all
```
Mainly for two reasons:

1. Simpler code - the alternative would include working out which links to add/delete (as happens currently), then iterating over the links to update their positions in the database. That makes the code more complicated and adds more iteration.

2. Less database I/O - deleting and batch importing guarantees two SQL statements for every operation (one `DELETE` and one `INSERT`), whereas the alternative would execute more statements in many cases (one `DELETE` or `INSERT` per change plus `UPDATE`s and `SELECT`s for updating link positions).

dougdroper replied:
I agree with the simpler code, although this way we lose the created_at timestamp (not sure that is an issue?)

rubenarakelyan replied:
Yup, that's one downside to this approach. I checked this with @tijmenb and the view is this shouldn't be a problem since we don't rely on it for anything on our side. Not sure if you use that field elsewhere.

MatMoore replied:
I agree with this approach but just wanted to note that there is still a separate insert for each link at the moment. So I'd expect this to perform roughly the same in the case where you're inserting a set of links for the first time, but if the user makes minor changes (adding a new tag to a long list) you lose the optimisation of just inserting the one you changed. 

That could maybe be remedied by inserting all the rows in one sql statement but I'm not sure you can do it with plain activerecord.

It might be worth checking for noops before doing anything, because publishing apps tend to send all their links when publishing a draft.

Have you run any of the benchmarks yet?

kevindew replied:
@MatMoore We (Publishing API) persuaded Ruben to drop the single insert to multiple ones so that's our fault 😉  We felt it was best to wait until there was performance issues.

@rubenarakelyan I've had a look at the above and I think you can save some a query on each group iteration by dropping the need to get `existing_content_ids`:

```ruby
          link_set.where(link_type: group).delete_all
          payload_content_ids.uniq.each_with_index do |content_id, i|
            links.create!(link_type: group, target_content_id: content_id, position: i)
          end
```

I also dug in a little to see if it would be difficult to go down the update and no-op route if nothing has changed and think this isn't too complex:

```ruby
          links = link_set.links.where(link_type: group)
          existing_content_ids = links.pluck(:target_content_id)
          payload_content_ids.uniq.each_with_index do |content_id, position|
            next unless content_id == existing_content_ids[postion]
            links.find_or_create_by!(target_content_id: content_id).update!(position: position)
          end
          to_delete = existing_content_ids - payload_content_ids
          links.where(target_content_id: to_delete).delete_all unless to_delete.empty?
```

dougdroper commented on app/commands/v2/patch_link_set.rb:
```diff
@@ -18,17 +18,25 @@ def call
           end
-
-          links.where(target_content_id: content_ids_to_delete).delete_all
+          Link.import links_to_create
```
I am not sure about using activerecord-import. Our average count of links per group is only 1.23, and in worst case (at the moment) there are 813 links, which I am sure the DB can handle quite quickly. Seems like adding another dependency a bit early?

dougdroper replied:
Also to @kevindew point about missing out on the Link validations.

rubenarakelyan replied:
I don't have any problem with reverting to the standard "one insert per row" if you don't think it will case any noticeable performance issues. Maybe we should benchmark before/after?

dougdroper replied:
Yes I agree, lets leave it out for now, then do some performance bench marks to see if this is an issue.

dougdroper commented:
@rubenarakelyan looks good to me now, can we get one or two tests just to cover the position logic. Also can we update the docs?

rubenarakelyan commented:
@dougdroper I've updated the `api.md` file and also added a specific test for making sure the order is correct. I also updated an existing order to use `eq` instead of `match_array` so that the ordering is also checked.

kevindew commented:
Can you update the ordering in the linkset model too: https://github.com/alphagov/publishing-api/blob/e4a5ad165a7917441b598e0817d155154e4a14b8/app/models/link_set.rb#L2 ?

dougdroper commented:
Is there any more work to do with the rendering to the content store? The expanded link set doesn't take into account any of the ordering (although it may be fine). Just wanted to know if that is part of this work.

rubenarakelyan commented:
@kevindew I've updated the `link_set` model to order by `link_type` and `position`. This is to keep links of the same type grouped together since the `position` is scoped to `link_set_id` then `link_type`.

@dougdroper Yes from discussing with @tijmenb I believe there is some work to do on the ordering of links when they are sent from the API to downstream consumers such as the Content Store. I think I'll need some more guidance on this though to work out what I need to change.

kevindew commented:
@rubenarakelyan great thanks for that.

Yeah I think you might have to dig into the [fun recursive query](https://github.com/alphagov/publishing-api/blob/master/app/presenters/queries/expanded_link_set.rb#L117) for the downstream ordering which does sound well suited to a separate PR.

kevindew commented:
Oh another thing, not really a concern for you @rubenarakelyan but more not for Publishing Platform team. 

I ran this migration locally and it took about 44seconds. Is this something we'll want to try deploy outside of a busy period? I think ALTER TABLE causes a write lock but I'm not certain.

boffbowsh commented on db/migrate/20161003145646_add_position_to_links.rb:
```diff
@@ -0,0 +1,5 @@
+class AddPositionToLinks < ActiveRecord::Migration
+  def change
+    add_column :links, :position, :integer, null: false, default: 0
```
The reason this takes so long is the default setting. It'd be better to do the following:

1. Add the column, allowing nulls with no default
2. `UPDATE "links" SET "position" = 0 WHERE "position" IS NULL;`
3. Alter the column and add the default

That way, PostgreSQL won't block at all.

kevindew replied:
ha oh dear - that was how it was wrote originally before I intervened.

Thanks @boffbowsh 👍  - Sorry @rubenarakelyan

rubenarakelyan replied:
No worries, I've changed this to do the migration in three steps.

boffbowsh commented on db/migrate/20161006115515_add_position_to_existing_links.rb:
```diff
@@ -0,0 +1,7 @@
+class AddPositionToExistingLinks < ActiveRecord::Migration
+  def change
+    Link.where("position IS NULL").each do |link|
```
```ruby
Link.where(position: nil).update_all(position: 0)
```

rubenarakelyan replied:
And it was changed
